src/server/roulette/rouletteSelection.ts

```diff
@@ -18,23 +18,37 @@ export function selectVariantUsingRoulette<V extends Variant, T extends Test<V>>
         return selectRandomVariant(test);
     }
 
-    const minWeight = 0.1; // Ensure no variant gets less than 10%
-    const variantsWithWeights: { weight: number; variantName: string }[] = testBanditData.variants
-        .map(({ variantName, mean }) => ({
+    // Ensure no variant gets less than 10%
+    const minWeight = 0.1;
+    if (testBanditData.variants.length > 9) {
+        return selectRandomVariant(test);
+    }
+
+    // First set weights only for variants that would fall below 10%
+    let sumOfNonReservedMeans = 0;
+    const minimumWeights = testBanditData.variants.map(({ variantName, mean }) => {
+        const weight = mean / sumOfMeans;
+        if (weight >= minWeight) {
+            sumOfNonReservedMeans += mean;
+        }
+        return {
             variantName,
-            weight: Math.max(mean / sumOfMeans, minWeight),
+            weight: weight < minWeight ? minWeight : 0,
+            mean,
+        };
+    });
+
+    // Now set the weights for the other variants
+    const remainder = 1 - minimumWeights.reduce((sum, v) => sum + v.weight, 0);
+    const weights: { weight: number; variantName: string }[] = minimumWeights
+        .map(({ variantName, weight, mean }) => ({
+            variantName,
+            weight: weight === 0 ? remainder * (mean / sumOfNonReservedMeans) : weight,
```

Comment:
If the mean is 0.01 and the sumOfNonReservedMeans is 1, then this would be given a weight of 0.01 -- much lower than a variant at 0 which gets 0.1. Assuming I'm following the code correctly.

Reply:
at this point we're setting the weight for any variant that we know would get more than minWeight (0.1), so this should be ok